lib/report.js

```diff
@@ -11,55 +11,152 @@ const line = (title, s) => {
+  for (const key in lines) {
+    if (!exclude.includes(key)) {
+      s += line(capitalize(key), lines[key]);
+    }
+  }
   return s;
 };
 
-module.exports = (metatests) => {
-
-  metatests.results = {
-    modules: 0,
-    targets: 0,
-    functions: 0,
-    values: 0,
-    classes: 0,
-    methods: 0,
-    tests: 0,
-    errors: 0,
-  };
-
-  const renderStack = stack => {
-    const path = process.cwd();
-    const lines = stack.split('\n');
-    lines.splice(1, 1);
-    const result = lines.map((line, i) => (
-      ' '.repeat(i ? 12 : 0) +
-      line.trim().replace(/at /g, '').replace(path, '')
-    )).filter(line => !line.includes('(internal/'));
-    return result.join('\n');
-  };
-
-  metatests.report = () => {
-    const show = () => {
-      const errCount = metatests.results.errors;
-      const successed = errCount === 0;
-      metatests.results['result'] = successed ? 'PASSED' : 'FAILED';
-      console.log(
-        'Test finished with following results' +
-        list(metatests.results)
-      );
-      if (!successed) process.exit(1);
+const renderStack = stack => {
+  const path = process.cwd();
+  const lines = stack.split('\n');
+  lines.splice(1, 1);
+  const result = lines.map((line, i) => (
+    ' '.repeat(i ? 12 : 0) +
+    line.trim().replace(/at /g, '').replace(path, '')
+  )).filter(line => !line.includes('(internal/'));
+  return result.join('\n');
+};
+
+class Reporter {
+
+  constructor(test = null) {
+    this.testsCount = 0;
+    this.test = test;
+    this._testDoneCallback = t => this.record(t);
+    if (test) this.addTest(test);
+  }
+
+  addTest(test) {
+    this.testsCount++;
+    test.on('done', this._testDoneCallback);
+  }
+
+  record(/* test */) {
+    this.testsCount--;
+    if (this.testsCount <= 0) {
+      process.nextTick(() => {
+        if (this.testsCount <= 0) this.report();
+      });
+    }
+  }
+
+  report() {
+    // ignore
+  }
+
+  runTests() {
+    if (this.test) this.test.runNow();
+  }
+}
+
+class SimpleReporter extends Reporter {
+
+  constructor(test = null) {
+    super(test);
+    this.failures = 0;
+    this.successes = 0;
+  }
+
+  record(test) {
+    super.record(test);
+    if (test.success) this.successes++;
+    else this.failures++;
+    console.log(`Test ${test.caption} finished` +
+      `${test.success ? 'successfully' : 'with failures'}`);
+  }
+
+  report() {
+    console.log(`Tests finished [+${this.successes} | -${this.failures}]`);
+  }
+}
+
+class SummaryReporter extends Reporter {
+
+  constructor(test) {
+    super(test);
+    this.results = {
+      // add one module to indicate any test that uses defaultReporter
+      modules: 1,
+      targets: 0,
+      functions: 0,
+      values: 0,
+      classes: 0,
+      methods: 0,
+      tests: 0,
+      errors: 0,
     };
-    metatests.run(show);
-  };
+  }
 
-  metatests.fail = (reason) => {
-    const message = ' Test failed: ' + reason.caption;
-    delete reason.caption;
+  _parseTestResults(test) {
```

Comment:
Why `_`?

Reply:
To indicate private-ness of the function - non-public api. Just to give a visual separation of public/'private' methods.